clang/lib/3C/ConstraintVariables.cpp

```diff
@@ -933,8 +963,53 @@ FunctionVariableConstraint::FunctionVariableConstraint(const Type *Ty,
   }
 
   // ConstraintVariable for the return
-  bool IsGeneric = FD != nullptr && getTypeVariableType(FD) != nullptr;
-  ReturnVar = new PVConstraint(RT, D, RETVAR, I, Ctx, &N, IsGeneric);
+  ReturnVar = allocateParamPair(RT, D, RETVAR, I, Ctx, &N, ReturnHasItype);
+}
+
+FunctionVariableConstraint::InternalExternalPair
+FunctionVariableConstraint::allocateParamPair(const clang::QualType &QT,
+                                              clang::DeclaratorDecl *D,
+                                              std::string N, ProgramInfo &I,
+                                              const clang::ASTContext &C,
+                                              std::string *InFunc,
+                                              bool VarAtomForChecked) {
+  bool IsGeneric = D && getTypeVariableType(D);
+  PVConstraint *PVExt = new PVConstraint(QT, D, N, I, C, InFunc, IsGeneric);
+  // For void pointers and function pointers, internal and external would need
+  // to be equated, so can we avoid allocating extra constraints.
+  if (QT->isVoidPointerType() || QT->isFunctionPointerType())
+    return {PVExt, PVExt};
+  PVConstraint *PVInt = new PVConstraint(QT, D, N, I, C, InFunc, IsGeneric,
+                                         VarAtomForChecked);
+  InternalExternalPair Pair = {PVInt, PVExt};
+  linkInternalExternalPair(I.getConstraints(), QT, N == RETVAR, Pair);
+  return Pair;
+}
+
+void FunctionVariableConstraint::linkInternalExternalPair(
+  Constraints &CS, QualType QT, bool IsReturn, InternalExternalPair Pair) {
+  assert(Pair.InternalConstraint->getCvars().size() ==
+         Pair.ExternalConstraint->getCvars().size());
+  assert(!(QT->isVoidPointerType() || QT->isFunctionPointerType()));
+  for (unsigned J = 0; J < Pair.InternalConstraint->getCvars().size(); J++) {
+    Atom *InternalA = Pair.InternalConstraint->getCvars()[J];
+    Atom *ExternalA = Pair.ExternalConstraint->getCvars()[J];
+    if (isa<VarAtom>(InternalA) || isa<VarAtom>(ExternalA)) {
+      // Equate pointer types for internal and external parameter constraint
+      // variables.
+      CS.addConstraint(CS.createGeq(InternalA, ExternalA, false));
+      CS.addConstraint(CS.createGeq(ExternalA, InternalA, false));
+      // Constrain Internal >= External. If external solves to wild, then so
+      // does the internal. Not that this doesn't mean any unsafe external
+      // use causes the internal variable to be wild because the external
+      // variable solves to WILD only when there is an unsafe use that
+      // cannot be resolved by inserting casts.
+      CS.addConstraint(CS.createGeq(InternalA, ExternalA, true));
+
+      if (!isa<ConstAtom>(ExternalA) && IsReturn && J > 0)
```

Comment:
What's different about returns?

Reply:
Looks like you are equating them (`=`) rather than flowing (`>=`) for ptyps ?

Reply:
But I'm confused how wild vs. ptyp constraints are being handled here. Probably best to talk in real-time, since this is the crux of the big change.